File: lib/go/hybrid.go
package radixip

import (
	"context"
	"encoding/json"
	"fmt"
	"net"
)

// HybridEngine acts as an orchestrator for a split-plane architecture.
// It routes writes to the controlPlane and reads to the dataPlane,
// synchronizing them via Redis.
type HybridEngine struct {
	controlPlane RadixEngine
	dataPlane    RadixEngine
	redis        *RedisClient
	channel      string
}

// NewHybridEngine creates a new HybridEngine
func NewHybridEngine(config *RadixConfig) (*HybridEngine, error) {
	if config == nil {
		return nil, fmt.Errorf("config is required for HybridEngine")
	}

	controlPlane := NewEngineWrapperWithTree(config.EngineVariant, config.NodeVariant, config.WriteCompressed)
	dataPlane := NewEngineWrapperWithTree(config.EngineVariant, config.NodeVariant, config.ReadCompressed)

	var redisClient *RedisClient
	var err error
	if config.Redis != nil {
		redisClient, err = NewRedisClient(*config.Redis)
		if err != nil {
			return nil, fmt.Errorf("failed to initialize redis client: %w", err)
		}
	}

	engine := &HybridEngine{
		controlPlane: controlPlane,
		dataPlane:    dataPlane,
		redis:        redisClient,
		channel:      config.RedisChannel,
	}

	// Boot-load the data plane from Redis if available
	if engine.redis != nil {
		ctx := context.Background()
		entries, err := engine.redis.HGetAll(ctx, "radixip:entries")
		if err == nil {
			for cidr, metaJSON := range entries {
				_, ipnet, err := net.ParseCIDR(cidr)
				if err == nil {
					var meta Metadata
					if json.Unmarshal([]byte(metaJSON), &meta) == nil {
						// Hydrate both planes
						engine.controlPlane.Insert(ipnet, meta)
						engine.dataPlane.Insert(ipnet, meta)
					}
				}
			}
		}
	}

	return engine, nil
}

// StartSync starts the Redis Pub/Sub synchronization loop for the data plane
func (e *HybridEngine) StartSync(ctx context.Context) error {
	if e.redis == nil {
		return fmt.Errorf("redis is not configured")
	}
	
	// Subscribe the dataPlane in a goroutine so it doesn't block
	go func() {
		err := e.redis.SubscribeEngineUpdates(ctx, e.channel, e.dataPlane)
		if err != nil {
			fmt.Printf("HybridEngine Redis sync stopped: %v\n", err)
		}
	}()
	
	return nil
}

// Insert adds a prefix with metadata to the control plane and publishes to Redis
func (e *HybridEngine) Insert(prefix *net.IPNet, metadata Metadata) error {
	// 1. Write to local control plane
	if err := e.controlPlane.Insert(prefix, metadata); err != nil {
		return err
	}

	// 2. Persist and Broadcast via Redis
	if e.redis != nil {
		ctx := context.Background()
		
		// Save state
		if jsonData, err := json.Marshal(metadata); err == nil {
			e.redis.HSet(ctx, "radixip:entries", prefix.String(), string(jsonData))
		}
		
		// Broadcast to all data planes (including our own local data plane)
		ipNetwork := IpNetwork{IP: prefix.IP, Mask: prefix.Mask}
		e.redis.PublishInsert(ctx, e.channel, ipNetwork, metadata)
	} else {
		// If no Redis, fallback to local sync
		e.dataPlane.Insert(prefix, metadata)
	}

	return nil
}

// Lookup queries the data plane exclusively
func (e *HybridEngine) Lookup(ip net.IP) *Metadata {
	return e.dataPlane.Lookup(ip)
}

// Remove deletes a prefix from the control plane and broadcasts to Redis
func (e *HybridEngine) Remove(prefix *net.IPNet) *Metadata {
	// 1. Remove from local control plane
	result := e.controlPlane.Remove(prefix)

	// 2. Remove and Broadcast via Redis
	if e.redis != nil {
		ctx := context.Background()
		e.redis.HDel(ctx, "radixip:entries", prefix.String())
		
		ipNetwork := IpNetwork{IP: prefix.IP, Mask: prefix.Mask}
		e.redis.PublishRemove(ctx, e.channel, ipNetwork)
	} else {
		// Fallback to local sync
		e.dataPlane.Remove(prefix)
	}

	return result
}

// Contains checks if a prefix exists in the data plane
func (e *HybridEngine) Contains(prefix *net.IPNet) bool {
	return e.dataPlane.Contains(prefix)
}

// Clear removes all entries and clears both planes
func (e *HybridEngine) Clear() {
	e.controlPlane.Clear()
	if e.redis != nil {
		ctx := context.Background()
		e.redis.PublishClear(ctx, e.channel)
		// We'd probably need a Lua script to clear the hash if we wanted to be perfectly clean
	} else {
		e.dataPlane.Clear()
	}
}

// Size returns the number of entries in the data plane
func (e *HybridEngine) Size() int {
	return e.dataPlane.Size()
}

// Stats aggregates statistics
func (e *HybridEngine) Stats() *EngineStats {
	// Return the stats of the data plane as it handles the reads
	return e.dataPlane.Stats()
}

File: lib/rust/src/hybrid.rs
use std::net::IpAddr;
#[cfg(feature = "redis")]
use std::sync::Arc;

use crate::config::RadixConfig;
use crate::engine::EngineWrapper;
#[cfg(feature = "redis")]
use crate::redis::RedisClient;
use crate::traits::RadixEngine;
use crate::types::{EngineStats, Metadata};
use ipnetwork::IpNetwork;

pub struct HybridEngine {
    control_plane: EngineWrapper,
    data_plane: EngineWrapper,
    #[cfg(feature = "redis")]
    redis: Option<Arc<RedisClient>>,
    channel: String,
}

impl HybridEngine {
    pub async fn new(config: &RadixConfig) -> Result<Self, String> {
        let control_plane = EngineWrapper::new(
            config.engine_variant.clone(),
            config.node_variant.clone(),
            config.write_compressed,
        );

        let data_plane = EngineWrapper::new(
            config.engine_variant.clone(),
            config.node_variant.clone(),
            config.read_compressed,
        );

        #[cfg(feature = "redis")]
        let redis = if let Some(redis_config) = &config.redis {
            Some(Arc::new(
                RedisClient::new(redis_config.clone())
                    .await
                    .map_err(|e| e.to_string())?,
            ))
        } else {
            None
        };

        let engine = Self {
            control_plane,
            data_plane,
            #[cfg(feature = "redis")]
            redis,
            channel: config.redis_channel.clone(),
        };

        #[cfg(feature = "redis")]
        if let Some(r) = &engine.redis {
            // Boot-load the data plane from Redis
            if let Ok(entries) = r.hgetall_sync("radixip:entries") {
                for (cidr, meta_json) in entries {
                    if let Ok(ipnet) = cidr.parse::<IpNetwork>() {
                        if let Ok(meta) = serde_json::from_str::<Metadata>(&meta_json) {
                            let _ = engine.control_plane.insert(ipnet.clone(), meta.clone());
                            let _ = engine.data_plane.insert(ipnet, meta);
                        }
                    }
                }
            }
        }

        Ok(engine)
    }

    #[cfg(feature = "redis")]
    pub async fn start_sync(&self) -> Result<(), String> {
        if let Some(redis) = &self.redis {
            let redis_clone = redis.clone();
            let channel = self.channel.clone();
            let data_plane = Arc::new(self.data_plane.clone());

            tokio::spawn(async move {
                if let Err(e) = redis_clone
                    .subscribe_engine_updates(&channel, data_plane)
                    .await
                {
                    eprintln!("HybridEngine Redis sync stopped: {}", e);
                }
            });
        }
        Ok(())
    }
}

impl RadixEngine for HybridEngine {
    fn insert(&self, prefix: IpNetwork, metadata: Metadata) -> Result<(), String> {
        self.control_plane
            .insert(prefix.clone(), metadata.clone())?;

        #[cfg(feature = "redis")]
        if let Some(redis) = &self.redis {
            if let Ok(json_data) = serde_json::to_string(&metadata) {
                let _ = redis.hset_sync("radixip:entries", &prefix.to_string(), &json_data);
            }
            let _ = redis.publish_insert(&self.channel, prefix, metadata);
        } else {
            let _ = self.data_plane.insert(prefix, metadata);
        }

        Ok(())
    }

    fn lookup(&self, ip: &IpAddr) -> Option<Metadata> {
        self.data_plane.lookup(ip)
    }

    fn remove(&self, prefix: &IpNetwork) -> Option<Metadata> {
        let result = self.control_plane.remove(prefix);

        #[cfg(feature = "redis")]
        if let Some(redis) = &self.redis {
            let _ = redis.hdel_sync("radixip:entries", &prefix.to_string());
            let _ = redis.publish_remove(&self.channel, prefix.clone());
        } else {
            let _ = self.data_plane.remove(prefix);
        }

        result
    }

    fn contains(&self, prefix: &IpNetwork) -> bool {
        self.data_plane.contains(prefix)
    }

    fn clear(&self) {
        self.control_plane.clear();
        #[cfg(feature = "redis")]
        if let Some(redis) = &self.redis {
            let _ = redis.publish_clear(&self.channel);
        } else {
            self.data_plane.clear();
        }
    }

    fn size(&self) -> usize {
        self.data_plane.size()
    }

    fn stats(&self) -> EngineStats {
        self.data_plane.stats()
    }
}
